cairo/ethereum/cancun/vm/precompiled_contracts/mapping.cairo

```diff
@@ -0,0 +1,102 @@
+from starkware.cairo.common.cairo_builtins import BitwiseBuiltin, KeccakBuiltin, PoseidonBuiltin
+from starkware.cairo.common.registers import get_label_location
+from starkware.cairo.common.math_cmp import is_le_felt
+from ethereum.cancun.vm import Evm
+from starkware.cairo.common.math import split_felt
+from ethereum.utils.numeric import divmod
+
+// currently 10 precompiles.
+const N_PRECOMPILES = 10;
+const HIGHEST_PRECOMPILE_LEADING_BYTE = 0x0a;
+
+// count 3 steps per index: precompile_address, call, precompile_fn
+const MAX_OFFSET = N_PRECOMPILES * 3;
+
+// @notice A table of (address, function ptr) for precompiled contracts
+// @dev Meant to be used with an `index` provided by the prover, that cannot be trusted.
+// @dev The address returned will need to be verified by the caller.
+// @param index The index of the address precompiled contract in the table (3 * number)
+// @param address The address of the precompiled contract called
+// @return (0, 0) if the address is not a precompile, otherwise (address, function_ptr)
+func precompile_table_lookup{range_check_ptr}(address: felt) -> (felt, felt) {
+    alloc_locals;
+
+    // Check if address is a valid precompile
+    // Addresses are little-endian, we want easy comparison to 0x1-0xa value, take the leading byte
+    let (address_high, address_low) = split_felt(address);
+    if (address_low != 0) {
+        return (0, 0);
+    }
+    let (leading_byte, _) = divmod(address_high, 2 ** (3 * 8));
+    if (leading_byte == 0) {
+        return (0, 0);
+    }
+    let addr_too_high = is_le_felt(HIGHEST_PRECOMPILE_LEADING_BYTE + 1, leading_byte);
+    if (addr_too_high != 0) {
+        return (0, 0);
+    }
+
+    // Provided the address is a valid precompile, get its index from the precompile table
+    // This enables non-sequential precompile addresses for the future...
+    let (local precompiled_contracts_location: felt*) = get_label_location(PRE_COMPILED_CONTRACTS);
+
+    tempvar index;
+    %{ precompile_index_from_address %}
+
+    let is_valid_index = is_le_felt(index, MAX_OFFSET);
+    with_attr error_message("precompile_table_lookup: index out of bounds") {
+        assert is_valid_index = 1;
+    }
+
+    // // at index: address, at index+1: call, at index+2: fn
+    let table_address = precompiled_contracts_location[index];
+    // Soundness: verify the fn we got from the jump table is the one associated with the address
```

Comment:
The check below is the verification that the address from the table == address requested right ?
The check for the fn associated is also necessary? If yes, why and how to implement it?

Reply:
>  address from the table == address requested right ?

exactly

> The check for the fn associated is also necessary?
No, because if the address at `index` is right, the function at `index + 2` is necessarily right as well